### PR TITLE
snap/pack, cmd/snap: allow specifying the filename of 'snap pack'

### PR DESCRIPTION
**Pull request by chipaca**
Before this change, 'snap pack' built the snap name from the yaml.
This is fiddly to use in automated tools in some situations; it's
better, there, to be able to inject the expected filename. This adds
that option, while preserving the old behaviour.

Cc @sergiusens.

### CODE REVIEW

File: snap/pack/pack.go [hunk 1]
```diff
@@ -119,9 +119,11 @@ func loadAndValidate(sourceDir string) (*snap.Info, error) {
...
 }
 
-func snapPath(info *snap.Info, targetDir string) string {
-	snapName := fmt.Sprintf("%s_%s_%v.snap", info.InstanceName(), info.Version, debArchitecture(info))
-	if targetDir != "" {
+func snapPath(info *snap.Info, targetDir, snapName string) string {
+	if snapName == "" {
+		snapName = fmt.Sprintf("%s_%s_%v.snap", info.InstanceName(), info.Version, debArchitecture(info))
+	}
+	if targetDir != "" && !filepath.IsAbs(snapName) {
```
Comment (stolowski) [inline snap/pack/pack.go]: Should we validate commands and errour out if absolute file path is given together with target path?

File: snap/pack/pack.go [hunk 2]
```diff
@@ -119,9 +119,11 @@ func loadAndValidate(sourceDir string) (*snap.Info, error) {
 	return info, nil
 }
 
-func snapPath(info *snap.Info, targetDir string) string {
-	snapName := fmt.Sprintf("%s_%s_%v.snap", info.InstanceName(), info.Version, debArchitecture(info))
-	if targetDir != "" {
+func snapPath(info *snap.Info, targetDir, snapName string) string {
+	if snapName == "" {
```
[the remark this answers is not in the document]
Comment (chipaca) [reply snap/pack/pack.go]: yeah, that's a bit weird. I'll fix it in a followup.

File: cmd/snap/cmd_pack.go [hunk 3]
```diff
@@ -52,15 +58,21 @@ messages.
...
+	cmd := addCommand("pack",
 		shortPackHelp,
 		longPackHelp,
 		func() flags.Commander {
 			return &packCmd{}
 		}, map[string]string{
 			// TRANSLATORS: This should not start with a lowercase letter.
 			"check-skeleton": i18n.G("Validate snap-dir metadata only"),
+			// TRANSLATORS: This should not start with a lowercase letter.
+			"filename": i18n.G("Use this filename"),
```
[the remark this answers is not in the document]
Comment (stolowski) [reply cmd/snap/cmd_pack.go]: Sounds good, thank you.

File: cmd/snap/cmd_pack.go [hunk 4]
```diff
@@ -52,18 +59,28 @@ messages.
...
+			"filename": i18n.G("Output to this filename"),
 		}, nil)
+	cmd.extra = func(cmd *flags.Command) {
+		// TRANSLATORS: this describes the default filename for a snap, e.g. core_16-2.35.2_amd64.snap
+		cmd.FindOptionByLongName("filename").DefaultMask = i18n.G("<name>_<version>_<architecture>.snap")
+	}
 }
 
 func (x *packCmd) Execute([]string) error {
+	if x.Positional.TargetDir != "" && x.Filename != "" && filepath.IsAbs(x.Filename) {
```
Comment (zyga) [inline cmd/snap/cmd_pack.go]: Nice touch :)

File: cmd/snap/main.go [hunk 5]
```diff
@@ -262,6 +263,9 @@ func Parser(cli *client.Client) *flags.Parser {
 			arg.Name = name
 			arg.Description = desc
 		}
+		if c.extra != nil {
```
Comment (zyga) [inline cmd/snap/main.go]: I think this and the struct warrants a comment.

Comment (chipaca) [reply cmd/snap/main.go]: I'll add it in a followup :-)

**Pull request merged**